### ISSUE TITLE
ResourceQuota causes high GameServer churn rate

### ISSUE THREAD

**Issue by nrwiersma** [googleforgames/agones#3384]
**What happened**:

We have a set of fleets in a namespace that is constrained by a `ResourceQuota`. The fleet tried to scale past what the `ResourceQuota` would allow. When creating new GameServers, they would go into an `Error` state as the `ResourceQuota` admission would forbid the creation of the Pod. This caused the `GameServerSet` to instantly delete the `GameServer` and create a new one, creating a very fast churn causing a lot of load on Kubernetes and etcd.

**What you expected to happen**:

When the `GameServer` hits an `Error` state, according to the state chart [for GameServers](https://agones.dev/site/docs/reference/gameserver/#gameserver-state-diagram) I would expect the GameServer not to be deleted, or for the `GameServerSet` controller to only recreate these GameServers at a fixed interval, to avoid high GameServer churn rate.

**How to reproduce it (as minimally and precisely as possible)**:

* Create a ResourceQuota with low `Hard` CPU or Memory limit
* Create a Fleet that will exceed those limits.

**Anything else we need to know?**:

**Environment**:
- Agones version: 1.34.0
- Kubernetes version (use `kubectl version`): v1.26.5-gke.270
- Cloud provider or hardware configuration: GKE
- Install method (yaml/helm): helm
- Troubleshooting guide log(s):
```
{"gs":{"metadata":{"name":"nick-test-kzt87-x4mpx","generateName":"nick-test-kzt87-","namespace":"test","uid":"7a980a9b-ddcb-416a-8bae-dc48a98537c3","resourceVersion":"32720772","generation":2,"creationTimestamp":"2023-09-18T14:04:30Z","labels":{"agones.dev/fleet":"nick-test","agones.dev/gameserverset":"nick-test-kzt87"},"annotations":{"agones.dev/sdk-version":"1.34.0"},"ownerReferences":[{"apiVersion":"agones.dev/v1","kind":"GameServerSet","name":"nick-test-kzt87","uid":"0ee9e358-0bff-42d2-9af2-c5a4322b6a28","controller":true,"blockOwnerDeletion":true}],"finalizers":["agones.dev"],"managedFields":[{"manager":"controller","operation":"Update","apiVersion":"agones.dev/v1","time":"2023-09-18T14:04:30Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:agones.dev/fleet":{},"f:agones.dev/gameserverset":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"0ee9e358-0bff-42d2-9af2-c5a4322b6a28\"}":{}}},"f:spec":{".":{},"f:container":{},"f:health":{},"f:immutableReplicas":{},"f:ports":{},"f:scheduling":{},"f:sdkServer":{},"f:template":{".":{},"f:metadata":{".":{},"f:annotations":{".":{},"f:kubectl.kubernetes.io/default-container":{}},"f:creationTimestamp":{},"f:labels":{".":{},"f:agones.dev/fleet":{}}},"f:spec":{".":{},"f:affinity":{".":{},"f:nodeAffinity":{".":{},"f:requiredDuringSchedulingIgnoredDuringExecution":{}}},"f:containers":{},"f:terminationGracePeriodSeconds":{},"f:tolerations":{}}}},"f:status":{".":{},"f:address":{},"f:addresses":{},"f:immutableReplicas":{},"f:nodeName":{},"f:players":{},"f:ports":{},"f:reservedUntil":{},"f:state":{}}}}]},"spec":{"container":"default","ports":[{"name":"http","portPolicy":"Dynamic","container":"default","containerPort":8080,"hostPort":17791,"protocol":"TCP"}],"health":{"periodSeconds":5,"failureThreshold":3,"initialDelaySeconds":5},"scheduling":"Packed","sdkServer":{"logLevel":"Info","grpcPort":9357,"httpPort":9358},"template":{"metadata":{"creationTimestamp":null,"labels":{"agones.dev/fleet":"nick-test"},"annotations":{"kubectl.kubernetes.io/default-container":"default"}},"spec":{"containers":[{"name":"default","image":"europe-docker.pkg.dev/test/dev/prod/testtest@sha256:4c323ab8c951bc63649e62ea95dec08736eef36ca388b1d7e9857fe2df008072","resources":{"requests":{"cpu":"250","memory":"100Gi"}},"imagePullPolicy":"IfNotPresent"}],"terminationGracePeriodSeconds":30,"affinity":{"nodeAffinity":{"requiredDuringSchedulingIgnoredDuringExecution":{"nodeSelectorTerms":[{"matchExpressions":[{"key":"game-server","operator":"Exists"}]}]}}},"tolerations":[{"key":"game-server","operator":"Equal","value":"true","effect":"NoExecute"}]}},"eviction":{"safe":"Never"}},"status":{"state":"Creating","ports":null,"address":"","addresses":null,"nodeName":"","reservedUntil":null,"players":null,"eviction":{"safe":"Never"}}},"gsKey":"test/nick-test-kzt87-x4mpx","message":"Pod created is forbidden","pod":{"metadata":{"creationTimestamp":null},"spec":{"containers":null},"status":{}},"severity":"error","source":"*gameservers.Controller","time":"2023-09-18T14:04:30.434315602Z"}
{"gs":{"metadata":{"name":"nick-test-kzt87-2dzzr","generateName":"nick-test-kzt87-","namespace":"test","uid":"d660c8c1-df28-4e4b-95ab-b97e25842458","resourceVersion":"32720784","generation":2,"creationTimestamp":"2023-09-18T14:04:30Z","labels":{"agones.dev/fleet":"nick-test","agones.dev/gameserverset":"nick-test-kzt87"},"annotations":{"agones.dev/sdk-version":"1.34.0"},"ownerReferences":[{"apiVersion":"agones.dev/v1","kind":"GameServerSet","name":"nick-test-kzt87","uid":"0ee9e358-0bff-42d2-9af2-c5a4322b6a28","controller":true,"blockOwnerDeletion":true}],"finalizers":["agones.dev"],"managedFields":[{"manager":"controller","operation":"Update","apiVersion":"agones.dev/v1","time":"2023-09-18T14:04:30Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:agones.dev/fleet":{},"f:agones.dev/gameserverset":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"0ee9e358-0bff-42d2-9af2-c5a4322b6a28\"}":{}}},"f:spec":{".":{},"f:container":{},"f:health":{},"f:immutableReplicas":{},"f:ports":{},"f:scheduling":{},"f:sdkServer":{},"f:template":{".":{},"f:metadata":{".":{},"f:annotations":{".":{},"f:kubectl.kubernetes.io/default-container":{}},"f:creationTimestamp":{},"f:labels":{".":{},"f:agones.dev/fleet":{}}},"f:spec":{".":{},"f:affinity":{".":{},"f:nodeAffinity":{".":{},"f:requiredDuringSchedulingIgnoredDuringExecution":{}}},"f:containers":{},"f:terminationGracePeriodSeconds":{},"f:tolerations":{}}}},"f:status":{".":{},"f:address":{},"f:addresses":{},"f:immutableReplicas":{},"f:nodeName":{},"f:players":{},"f:ports":{},"f:reservedUntil":{},"f:state":{}}}}]},"spec":{"container":"default","ports":[{"name":"http","portPolicy":"Dynamic","container":"default","containerPort":8080,"hostPort":22653,"protocol":"TCP"}],"health":{"periodSeconds":5,"failureThreshold":3,"initialDelaySeconds":5},"scheduling":"Packed","sdkServer":{"logLevel":"Info","grpcPort":9357,"httpPort":9358},"template":{"metadata":{"creationTimestamp":null,"labels":{"agones.dev/fleet":"nick-test"},"annotations":{"kubectl.kubernetes.io/default-container":"default"}},"spec":{"containers":[{"name":"default","image":"europe-docker.pkg.dev/test/dev/prod/testtest@sha256:4c323ab8c951bc63649e62ea95dec08736eef36ca388b1d7e9857fe2df008072","resources":{"requests":{"cpu":"250","memory":"100Gi"}},"imagePullPolicy":"IfNotPresent"}],"terminationGracePeriodSeconds":30,"affinity":{"nodeAffinity":{"requiredDuringSchedulingIgnoredDuringExecution":{"nodeSelectorTerms":[{"matchExpressions":[{"key":"game-server","operator":"Exists"}]}]}}},"tolerations":[{"key":"game-server","operator":"Equal","value":"true","effect":"NoExecute"}]}},"eviction":{"safe":"Never"}},"status":{"state":"Creating","ports":null,"address":"","addresses":null,"nodeName":"","reservedUntil":null,"players":null,"eviction":{"safe":"Never"}}},"gsKey":"test/nick-test-kzt87-2dzzr","message":"Pod created is forbidden","pod":{"metadata":{"creationTimestamp":null},"spec":{"containers":null},"status":{}},"severity":"error","source":"*gameservers.Controller","time":"2023-09-18T14:04:30.445950984Z"}
{"gs":{"metadata":{"name":"nick-test-cpl26-vx66f","generateName":"nick-test-cpl26-","namespace":"test","uid":"9658d3fa-f2e2-4248-863a-a76e39942bd7","resourceVersion":"32720787","generation":2,"creationTimestamp":"2023-09-18T14:04:30Z","labels":{"agones.dev/fleet":"nick-test","agones.dev/gameserverset":"nick-test-cpl26"},"annotations":{"agones.dev/sdk-version":"1.34.0"},"ownerReferences":[{"apiVersion":"agones.dev/v1","kind":"GameServerSet","name":"nick-test-cpl26","uid":"4acb6550-e3dd-4178-8bfc-ce2b759780d6","controller":true,"blockOwnerDeletion":true}],"finalizers":["agones.dev"],"managedFields":[{"manager":"controller","operation":"Update","apiVersion":"agones.dev/v1","time":"2023-09-18T14:04:30Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:generateName":{},"f:labels":{".":{},"f:agones.dev/fleet":{},"f:agones.dev/gameserverset":{}},"f:ownerReferences":{".":{},"k:{\"uid\":\"4acb6550-e3dd-4178-8bfc-ce2b759780d6\"}":{}}},"f:spec":{".":{},"f:container":{},"f:health":{},"f:immutableReplicas":{},"f:ports":{},"f:scheduling":{},"f:sdkServer":{},"f:template":{".":{},"f:metadata":{".":{},"f:annotations":{".":{},"f:kubectl.kubernetes.io/default-container":{}},"f:creationTimestamp":{},"f:labels":{".":{},"f:agones.dev/fleet":{}}},"f:spec":{".":{},"f:affinity":{".":{},"f:nodeAffinity":{".":{},"f:requiredDuringSchedulingIgnoredDuringExecution":{}}},"f:containers":{},"f:terminationGracePeriodSeconds":{},"f:tolerations":{}}}},"f:status":{".":{},"f:address":{},"f:addresses":{},"f:immutableReplicas":{},"f:nodeName":{},"f:players":{},"f:ports":{},"f:reservedUntil":{},"f:state":{}}}}]},"spec":{"container":"default","ports":[{"name":"http","portPolicy":"Dynamic","container":"default","containerPort":8080,"hostPort":17731,"protocol":"TCP"}],"health":{"periodSeconds":5,"failureThreshold":3,"initialDelaySeconds":5},"scheduling":"Packed","sdkServer":{"logLevel":"Info","grpcPort":9357,"httpPort":9358},"template":{"metadata":{"creationTimestamp":null,"labels":{"agones.dev/fleet":"nick-test"},"annotations":{"kubectl.kubernetes.io/default-container":"default"}},"spec":{"containers":[{"name":"default","image":"europe-docker.pkg.dev/test/dev/prod/testtest@sha256:4c323ab8c951bc63649e62ea95dec08736eef36ca388b1d7e9857fe2df008072","resources":{"requests":{"cpu":"2","memory":"100Mi"}},"imagePullPolicy":"IfNotPresent"}],"terminationGracePeriodSeconds":30,"affinity":{"nodeAffinity":{"requiredDuringSchedulingIgnoredDuringExecution":{"nodeSelectorTerms":[{"matchExpressions":[{"key":"game-server","operator":"Exists"}]}]}}},"tolerations":[{"key":"game-server","operator":"Equal","value":"true","effect":"NoExecute"}]}},"eviction":{"safe":"Never"}},"status":{"state":"Creating","ports":null,"address":"","addresses":null,"nodeName":"","reservedUntil":null,"players":null,"eviction":{"safe":"Never"}}},"gsKey":"test/nick-test-cpl26-vx66f","message":"Pod created is forbidden","pod":{"metadata":{"creationTimestamp":null},"spec":{"containers":null},"status":{}},"severity":"error","source":"*gameservers.Controller","time":"2023-09-18T14:04:30.463975513Z"}
```
- Others:

**Comment by nrwiersma** [comment]
All fair points, if this is the intended behaviour. Perhaps a note in the docs to point out that it doesn't play well ResourceQuotas could save someone else an unexpected production incident?

However, from my understanding of the code which may be incomplete, the occasions for an Error state are quiet limited (Invalid and Forbidden) and so could have different behaviour to an Unhealthy state, but this is an off the top of my head thought.

**Comment by markmandel** [comment]
> However, from my understanding of the code which may be incomplete, the occasions for an Error state are quiet limited (Invalid and Forbidden) and so could have different behaviour to an Unhealthy state, but this is an off the top of my head thought.

No that's a good point. It can happen if something truly weird happens when creating the Pod definition from a GameServer, but that should only happen in a truly broken scenario, but yes those are the two main scenarios.

I do think we still want to delete the GameServer in question (maybe you realise what's going on, and increase the ResourceQuota, and therefore would like to replace the GameServer), but ideally it shouldn't happen so fast that it causes an outage.

The other potential benefit here is that if you make an invalid Pod in your Fleet, you might have a moment to actually see it happen.

The interesting question is.... how do we do that?

The only thought I have (and we would need to test this, is here:
https://github.com/googleforgames/agones/blob/18163e820cc567c13fd71ae444e63b6986042c53/pkg/gameserversets/controller.go#L493-L509

Basically put a `time.Sleep(5 * time.Second)` in front of `_, err := c.gameServerGetter.GameServers(gs.Namespace).Update(ctx, gsCopy, metav1.UpdateOptions{})` - although I'm not 100% sure what a delay like that will do to the rest of the system.

(Also a Fleet Autoscaler is going to autoscale - I don't think we can totally prevent you from potential footguns here, but we can at least help a little)

**Comment by markmandel** [comment]
@zmerlynn @roberthbailey WDYT? scary idea, or could possibly work?

**Comment by markmandel** [comment]
Another potential idea, which is less scary (because it's not slowing down a working queue), is to put it here:

https://github.com/googleforgames/agones/blob/18163e820cc567c13fd71ae444e63b6986042c53/pkg/gameserversets/controller.go#L258-L279

Wherein the code checks the GameServer itself, and if the GameServer is in `Error` state, rather than queuing the GameServerSet immediately, could enqueue in 30s (or 5 or 10? not sure what the right number is).

Downside there though -- if there is a healthy GameServer that needs replacing right away, that's not going to happen until after the GameServerSet has waiting it's allotted delay time.... what makes me think this is bad idea (but leaving it here just in case people find it has merit in a way I'm not considering).

**Comment by nrwiersma** [comment]
My thought, although I dont have a great overview of the full consequence, was to handle it here:

https://github.com/googleforgames/agones/blob/c93f709a14dbc3fa601f428a2f07e1a662c2dc80/pkg/gameserversets/controller.go#L431-L434

It might be possible to count the Error state game servers as up for a short period and not schedule for deletion immediately. Basically if you can work out how long a GameServer has been in a state, you could consider it up for 30s, and in the next cycle after that, schedule deletion and replace it. This would potentially slow Error states down, but nothing else.

**Comment by markmandel** [comment]
> Basically if you can work out how long a GameServer has been in a state, you could consider it up for 30s

🤔  that could work - we'd have to add an annotation to the GameServer for the timestamp for when the GameServer was moved to `Error` state, but we can do that in `moveToError`.

That way you don't get a delay in the subsequent processing of the `GameServerSet` - that works!

By the same token, you could do the same logic in:

https://github.com/googleforgames/agones/blob/c93f709a14dbc3fa601f428a2f07e1a662c2dc80/pkg/gameserversets/controller.go#L499

Such that if the time hasn't passed, then don't delete the `GameServer` yet, and then process it again next time around - basically don't _not_ add it to the deletion queue, but handle the check in the deletion logic. Might be a bit cleaner that way.

But the idea makes sense!
